### ISSUE TITLE
Error: toNetInput - Problem with canvas.loadImage() + faceapi.detectAllFaces()

### ISSUE THREAD

**Issue by Gabe-Soares** [justadudewhohacks/face-api.js#729]
Hey!

My name is Gabriel and I'm using face-api.js to apply FaceRecognition on my final essay, on my way to graduate.

With the objective of trying the face-api and understand how it works, I created the following test.js:
```import '@tensorflow/tfjs-node';
import faceapi from "face-api.js";

import * as canvas from 'canvas';
const { Canvas, Image, ImageData } = canvas;
faceapi.env.monkeyPatch({ Canvas, Image, ImageData });

const faceDetectionNet = faceapi.nets.ssdMobilenetv1;
const minConfidence = 0.5;
const faceDetectionOptions = new faceapi.SsdMobilenetv1Options({ minConfidence });

async function runRecognition() {    
    await faceDetectionNet.loadFromDisk('./weights');
    await faceapi.nets.faceLandmark68Net.loadFromDisk('./weights');
    await faceapi.nets.faceRecognitionNet.loadFromDisk('./weights');

    const REFERENCE_IMAGE = './bbt1.jpg';
    const QUERY_IMAGE = './bbt2.jpg';

    const referenceImage = await canvas.default.loadImage(REFERENCE_IMAGE);
    const queryImage = await canvas.default.loadImage(QUERY_IMAGE);

    const detections_ref = await faceapi.detectAllFaces(referenceImage, faceDetectionOptions)
        .withFaceLandmarks()
        .withFaceDescriptors();

    const detections_query = await faceapi.detectAllFaces(queryImage, faceDetectionOptions)
        .withFaceLandmarks()
        .withFaceDescriptors();

    const faceMatcher = new faceapi.FaceMatcher(detections_ref);

    detections_query.forEach(fd => {
        const bestMatch = faceMatcher.findBestMatch(fd.descriptor);
        console.log(bestMatch.toString());
    });
}

runRecognition();
```

But I wasn't able to make it work, right now, I'm getting the following error:
"UnhandledPromiseRejectionWarning: Error: toNetInput - expected media to be of type HTMLImageElement | HTMLVideoElement | HTMLCanvasElement | tf.Tensor3D, or to be an element id
    at C:\Users\gabri\Desktop\TCC\face-api-teste\node_modules\face-api.js\build\commonjs\dom\toNetInput.js:38:35"

Does anyone have any idea of what the problem is?

Thanks for the help 😄

**Comment by vladmandic** [comment]
what is the result and type of `const referenceImage = await canvas.default.loadImage(REFERENCE_IMAGE);` ?

if you're not drawing on canvas in nodejs, you don't actually need 3rd party libraries and monkey patches just to load image and use it in `face-api` - you can use native `tfjs-node` methods directly, take a look at <https://github.com/vladmandic/face-api/blob/master/example/node.js>, specifically 'image()' method.

**Comment by Gabe-Soares** [comment]
Thanks for the tip, I'll try it as in your example as soon as I can, and then I comment here if it worked.

About the result of `const referenceImage = await canvas.default.loadImage(REFERENCE_IMAGE);`, I logged the object and it printed:
[Image:1366x768 ./bbt1.jpg complete]

**Comment by vladmandic** [comment]
that doesn't look like `Image` type, that looks like something canvas internal.
`face-api` perfoms check:   `input instanceof Image || input instanceof Canvas || input instanceof Video` and if that fails, you get your error `expected media to be of type ...`

took a quick look at canvas and their example shows how to create canvas and then load image into it and not how to use `canvas.default` to create image type directly.

```js
const referenceImage = await canvas.loadImage(REFERENCE_IMAGE);
const myCanvas = canvas.createCanvas(200, 200)
const ctx = myCanvas.getContext('2d')
ctx.drawImage(image, 0, 0, 200, 200)
```
and now you have `myCanvas` that you can pass to face-api.

but like i said, you don't need canvas at all.

**Comment by Gabe-Soares** [comment]
I'm sorry for taking too long to get back here with news!
But thanks a lot @vladmandic, you saved me and my group! hahaha
You were right, I refactored my code to something more alike your example, and it worked nice and smooth!
I'll set this issue as solved, but only for purpose of sharing knowledge, could you please explain to me when would I need or not need canvas?

Thanks again!

**Comment by vladmandic** [comment]
@Gabe-Soares 

if you want to draw results on top of an image (and then you'd probably also want to save that image to disk) - then you need a canvas.
but if you're just displaying results as text, no need for canvas as everything can be done without it.

**Comment by juancruzsosagab** [comment]
Hello @Gabe-Soares , im having de same issue, could you share what did you do? thanks a lot!

**Comment by eze-villegas** [comment]
Hi @juancruzsosagab, were you able to make any progress with this problem?
I am stuck on the same thing.

**Comment by vladmandic** [comment]
@juancruzsosagab @eze-villegas this issue was successfully resolved which you can see if you read through it. and posting "me too" on an issue that was closed 1.5 years ago does not help. if you have a problem, document it and open a new issue.

**Comment by mmm1831** [comment]
An old error description, maybe it will help someone.

I process many images in a directory and the error occurs when the image contains an **animated** GIF

```
Environment:
    face-api with NodeJS

Code:
    const options = new faceapi.SsdMobilenetv1Options({ minConfidence: 0.09, maxResults: 10 });
    const buffer = fs.readFileSync(fullpath);
    const decodeT = faceapi.tf.node.decodeImage(buffer, 3);
    const expandT = faceapi.tf.expandDims(decodeT, 0);
    const results = await faceapi.detectAllFaces(expandT, options) <= ERROR with animated GIF


Error: 
    toNetInput - expected media to be of type HTMLImageElement | HTMLVideoElement | HTMLCanvasElement | tf.Tensor3D, or to be an element id
    at /home/osboxes/dev/apps/server/node_modules/@vladmandic/face-api/dist/face-api.node.js:7:25421
```
